### ISSUE TITLE
In some cases no inheritance diagram is drawn.

### ISSUE THREAD

**Issue by Paebbels** [sphinx-doc/sphinx#3164]
I'm reviewing the automatically generated documentation and found that sometime the inheritance diagram is missing:
![image](https://cloud.githubusercontent.com/assets/956109/20460694/b6391730-aeeb-11e6-83f4-c66e1726065d.png)

The base class `Base.Shared.Shared` is imported via: `from Base.Shared import Shared` and documented in the same way. It has a inheritance diagram:
![image](https://cloud.githubusercontent.com/assets/956109/20460730/ce816df4-aeed-11e6-868b-e8b25150fbcc.png)

I noticed that autoapi has several problems when names are used for modules as well as packages or classes. **This this the same case with autodoc?**

The base class `Base.Shared.Shared` is defined in the same package (`Base`). How can I investigate this issue to write a better bug report for Sphinx?

```Python
# ...
from Base.Shared import Shared
# ....

class Compiler(Shared):
	"""Base class for all Compiler classes."""
	
	_ENVIRONMENT =    Environment.Synthesis
	_vhdlVersion =    VHDLVersion.VHDL93
	
	class __Directories__(Shared.__Directories__):
		Netlist =     None
		Source =      None
		Destination = None
	
	@DocumentMemberAttribute()
	def __init__(self, host, dryRun, noCleanUp):
		"""Class initializer
		
		:type  host:      object
		:param host:      The hosting instance for this instance.
		:type  dryRun:    bool
		:param dryRun:    Enable dry-run mode
		:type  noCleanUp: bool
		:param noCleanUp: Don't clean up after a run.
		"""
		super().__init__(host, dryRun)
		# ...
```

----------------------------
I can supply the complete documentation incl. Python code and loaded extensions as a branch in GitHub. If you decide to debug it by yourself. But it's quite huge :).

**Comment by tk0miya** [comment]
Both autodoc and inheritance_diagram use `sphinx.pycode` module to analyze the python codes.
So it might been related.

Anyway, to investigate your problem, we have to reproduce the problem at first. So could you share your documentation. Of course, it would be nice if there is any small reproducible project :-)

**Comment by Paebbels** [comment]
As I saw, the current online version of the documentation shows already this behavior:

Compiler class: [Base.Compiler.Compiler](http://poc-library.readthedocs.io/en/release/PyInfrastructure/Base.Compiler.html#Base.Compiler.Compiler)
Shared class: [Base.Shared.Shared](http://poc-library.readthedocs.io/en/release/PyInfrastructure/Base.Shared.html#Base.Shared.Shared)

The repository is: https://github.com/VLSI-EDA/PoC (Please use the 'release' branch.)

The documentation is located in [`.\docs`](https://github.com/VLSI-EDA/PoC/tree/release/docs?ts=2) and the Python code in [`.\py`](https://github.com/VLSI-EDA/PoC/tree/release/py?ts=2). `.\py\PoC.py` is the main program and is wrapped in a Bash or PowerShell script. But I don't think you need this information to investigate the documentation issue :). The ArgParse documentation is generated by `autoprogram`, which uses a special Python entry point called [`.\docs\PoCSphinx.py`](https://github.com/VLSI-EDA/PoC/blob/release/docs/PoCSphinx.py?ts=2).

The documentation uses 7 autoapi directives - one per package (Base, Compiler, ...) - and one autoapi directive for the main program `.\py\PoC.py`.

I think I can provide a new documentation version in circa one day, which eliminates some linking bugs doe to double used names. But this should effect the reported issue, because I see this behavior in the online and offline version.

I you think it might help or improve the debugging, I'm also available via Gitter chat or Google Hangouts.

**Comment by Paebbels** [comment]
I just found these warnings/errors:

```
D:\git\PoC\docs\PyInfrastructure\Compiler.ISECompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.ISECompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.LSECompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.LSECompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.QuartusCompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.QuartusCompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.VivadoCompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.VivadoCompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.XCICompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.XCICompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.XCOCompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.XCOCompiler.rst:17: ERROR: Document or section may not begin with a transition.
D:\git\PoC\docs\PyInfrastructure\Compiler.XSTCompiler.rst:7: WARNING: No classes found for inheritance diagram
D:\git\PoC\docs\PyInfrastructure\Compiler.XSTCompiler.rst:17: ERROR: Document or section may not begin with a transition.
```

These classes are inheriting again from `Compiler`, but where processed in another autoapi run.

What does:

> Document or section may not begin with a transition.

mean?

**Comment by tk0miya** [comment]
Thanks, I reproduced. I will take a look later.

> Document or section may not begin with a transition.

transition is a element of docutils like separator.
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#transitions

The message says the first element of document or section may not be a transition.

**Comment by Paebbels** [comment]
Thanks, I found the "transition" problem. I altered the autoapi template and introduced lines. If some sections where not generated, lone standing transitions could appear... => fixed.

I also found this warning:

> D:\git\PoC\docs\PyInfrastructure\Base.Compiler.rst:9: WARNING: No classes found for inheritance diagram

... which matches the missing inheritance diagram. This file contains the class `Base.Compiler.Compiler`, which has no inheritance diagram.

**Comment by tk0miya** [comment]
Finally I found this is a bug of `sphinx.ext.inheritance_diagram`.

In your case, reST is marked up like following:
```
.. currentmodule:: Base.Compiler
.. inheritance-diagram:: Compiler
```
The `inherintance-diagram` searches the class definitions with following order:
1. `Compiler`
2. `Base.Compiler.Compiler`

As a result, the directive ignores `currentmodule` settings if there is `Compile` module on root namespace.

>D:\git\PoC\docs\PyInfrastructure\Base.Compiler.rst:9: WARNING: No classes found for inheritance diagram

This message says `inheritance-diagram` loaded `Compiler` module (not `Base.Compiler` module  and `Base.Compiler.Compiler` class), But there are no classes.

**Comment by tk0miya** [comment]
I put #3171 to fix this.
I will merge this tomorrow.
Please comment if you still have problem with this.

**Comment by Paebbels** [comment]
Is there any easy way to test it?
Currently all packages are controlled by PIP.
I could:
* overwrite the file in my site-package folder with your new version
* copy this file into my `_extensions` folder and use `inheritance_diagram` instead of `sphinx.ext.inheritance_diagram`

right?

**Comment by tk0miya** [comment]
Please install it using following command:
```
pip install https://github.com/tk0miya/sphinx/archive/3164_fix_inheritance_diagram.zip
```

**Comment by Paebbels** [comment]
So the inheritance diagram is drawn and looks like this:
![image](https://cloud.githubusercontent.com/assets/956109/20528798/e409ae2a-b0cd-11e6-8d5e-72251134543f.png)

I noticed two things:
1. If the diagram gets bigger, it should have get a scrollbar. I'm using another Sphinx extension called wavedrom, which solved this job by adding a &lt;div style="overflow-x:auto"&gt; - box.
   
   ![image](https://cloud.githubusercontent.com/assets/956109/20528871/309ee44e-b0ce-11e6-81cf-66d35e97d640.png)

2. The inheritance diagram used to be clickable, but since I installed your update, it can't be clicked anymore. I also noticed that the names in the list of classes are not clickable. I marked the formerly clickable elements in blue. Are the references broken?
   ![image](https://cloud.githubusercontent.com/assets/956109/20528988/aae8a384-b0ce-11e6-950a-e5eba3e6f200.png)

**Comment by tk0miya** [comment]
Thank you for quick reviewing.

The first problem is not related with this issue.
Could you file it as new issue please?

I will take a look the second one later.
But I fixed `sphinx.ext.inheritance_diagram` only in the PR.
So the case of the list of classes might be not related with it.

**Comment by Paebbels** [comment]
The first problem is filed as #3172.

**Comment by tk0miya** [comment]
Thanks!

**Comment by tk0miya** [comment]
>The inheritance diagram used to be clickable, but since I installed your update, it can't be clicked anymore.

It seems your documents on RTD already can't be clicked.
The problem really came from my PR?
http://poc-library.readthedocs.io/en/release/PyInfrastructure/Base.Compiler.html

> I also noticed that the names in the list of classes are not clickable. I marked the formerly clickable elements in blue. Are the references broken?

This is same as above.
Current your documentation is already not clickable.
http://poc-library.readthedocs.io/en/release/PyInfrastructure/Base.Simulator.html

It seems some python objects are defined as `:noindex:`
```
.. autoexception:: SimulatorException
   :members:
   :noindex:
   :private-members:
   :inherited-members:
   :undoc-members:

   .. rubric:: Inheritance
   .. inheritance-diagram:: SimulatorException

   .. rubric:: Members
```
Actually, there are only a very few items in your python module index.
http://poc-library.readthedocs.io/en/release/py-modindex.html

Could you investigate carefully before reporting please?
It seems your project is already too complicated.
We can't determine it is a bug of Sphinx or not because of the 3rd party extensions you used.
Please try it with bare Sphinx before reporting, or doubt your extensions.

Anyway, it seems your problem first reported is resolved.
Please file a new issue if you still have a problem.

**Comment by tk0miya** [comment]
Now I merged #3171.
It will be released as Sphinx-1.4.9
Thank you for reporting

**Comment by Paebbels** [comment]
Thank you for fixing this reported issue.
I added `:noindex:` because the Python documentation is only a subpart of the complete documentation and I didn't want the Python stuff to pollute the index, so I set this option. I'll investigate when this behavior changed and try a smaller example.

So just one last question to understand your note:
Are the inheritance diagram links based on indexing or on labels?
Is this the same case for the Python module index?
